### PR TITLE
RL- add MDP, val-iter, and first draft index

### PR DESCRIPTION
**Pull request by rasoolfa**
Added MDP, value-iteration, and index notebooks. Als, updated d2l/torch.py and d2l.bib

By submitting this pull request, I confirm that you can use, modify,
copy, and redistribute this contribution, under the terms of your
choice.

### CODE REVIEW

File: chapter_reinforcement-learning/value-iter.md [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+# now set up the environment
+env_info = d2l.make_env('FrozenLake-v1',seed=seed)
+```
+
+In the FrozenLake environment, the robot moves on a $4 \times 4$ grid (these are the states) with actions that are "up" ($\uparrow$), "down" ($\rightarrow$), "left" ($\leftarrow$), and "right" ($\rightarrow$). The environment contains a number of holes (H) cells and frozen (F) cells as well as a goal cell (G), all of which are unknown to the robot. To keep the problem simple, we assume the robot has reliable actions, i.e. $P(s' \mid s, a) = 1$ for all $s \in \mathcal{S}, a \in \mathcal{A}$. If the robot reaches the goal, the trial ends and the robot receives a reward of $1$ irrespective of the action; the reward at any other state is $0$ for all actions. The objective of the robot is to learn a policy that reaches the goal location (G) from a given start location (S) (this is $s_0$) to maximize the *return*.
+
+The following function implements Value Iteration:
+
+```{.python .input}
+#@tab all
```
Comment (AnirudhDagar) [inline chapter_reinforcement-learning/value-iter.md:118]: This is old syntax not used anymore.

```suggestion
%%tab all
```

File: chapter_reinforcement-learning/value-iter.md [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+$$
+This algorithm is known as policy evaluation and is useful to compute the value function given the policy. Again, it turns out that as $k \to \infty$ these updates converge to the correct value function irrespective of the initialization $V_0$,
+$$V^\pi(s) = \lim_{k \to \infty} V^\pi_k(s);\ \text{for all states } s \in \mathcal{S}.$$
+The algorithm for computing the action-value function $Q^\pi(s, a)$ of a policy $\pi$ is analogous.
+
+## Implementation of Value Iteration
+:label:`subsec_valueitercode`
+We next show how to implement Value Iteration for a navigation problem called FrozenLake from [Open AI Gym](https://gym.openai.com). We first need to setup the enviroment as shown in the following code.
+```{.python .input}
+#@tab all
```
Comment (AnirudhDagar) [inline chapter_reinforcement-learning/value-iter.md:95]:
```suggestion
%%tab all
```

**Review by AnirudhDagar — COMMENTED** [review]
Hi, I made a cursory pass on the code and suggest some style changes for consistency throughout the book. 

Other than the smaller style nits, we should add the notebooks which save the functions in d2l.

File: chapter_reinforcement-learning/value-iter.md [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+%matplotlib inline
+%pip install gym==0.21.0
+import numpy as np
+import random
+from d2l import torch as d2l
+
+seed = 0       # Random number generator seed
+gamma = 0.95   # Discount factor
+num_iters = 10 # Number of iterations
+random.seed(seed) # Set the random seed to ensure results can be reproduced
```
[the remark this answers is not in the document]
Comment (astonzhang) [reply chapter_reinforcement-learning/value-iter.md]: @rasoolfa, I've committed this change

File: chapter_reinforcement-learning/value-iter.md [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+np.random.seed(seed)
+
+# now set up the environment
+env_info = d2l.make_env('FrozenLake-v1',seed=seed)
```
[the remark this answers is not in the document]
Comment (astonzhang) [reply chapter_reinforcement-learning/value-iter.md:110]: @rasoolfa, I've committed this change

File: chapter_reinforcement-learning/value-iter.md [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+        num_iters: number of iteration
+        env_info: contains MDP and environment related information
+    '''
+    env_desc      = env_info['desc'] # 2D array shows what each item means
+    prob_idx      = env_info['trans_prob_idx']
+    nextstate_idx = env_info['nextstate_idx']
+    reward_idx    = env_info['reward_idx']
+    num_states    = env_info['num_states']
+    num_actions   = env_info['num_actions']
+    mdp           = env_info['mdp']
```
[the remark this answers is not in the document]
Comment (astonzhang) [reply chapter_reinforcement-learning/value-iter.md:132]: @rasoolfa, I've committed this change

File: chapter_reinforcement-learning/value-iter.md (hunk 1 above)
Comment (astonzhang) [reply chapter_reinforcement-learning/value-iter.md:118]: @rasoolfa, I've committed this change

File: chapter_reinforcement-learning/value-iter.md (hunk 2 above)
Comment (astonzhang) [reply chapter_reinforcement-learning/value-iter.md:95]: @rasoolfa, I've committed this change

File: chapter_reinforcement-learning/value-iter.md [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+The algorithm for computing the action-value function $Q^\pi(s, a)$ of a policy $\pi$ is analogous.
+
+## Implementation of Value Iteration
+:label:`subsec_valueitercode`
+We next show how to implement Value Iteration for a navigation problem called FrozenLake from [Open AI Gym](https://gym.openai.com). We first need to setup the enviroment as shown in the following code.
+```{.python .input}
+#@tab all
+
+%matplotlib inline
+%pip install gym==0.21.0
```
[the remark this answers is not in the document]
Comment (astonzhang) [reply chapter_reinforcement-learning/value-iter.md:98]: +1
We'll remove it since it's in https://github.com/rasoolfa/d2l-en/blob/vol2-rl/setup.py#L11

File: chapter_reinforcement-learning/value-iter.md [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+The algorithm for computing the action-value function $Q^\pi(s, a)$ of a policy $\pi$ is analogous.
+
+## Implementation of Value Iteration
+:label:`subsec_valueitercode`
+We next show how to implement Value Iteration for a navigation problem called FrozenLake from [Open AI Gym](https://gym.openai.com). We first need to setup the enviroment as shown in the following code.
+```{.python .input}
+%%tab all
+
+%matplotlib inline
+%pip install gym==0.21.0
```
Comment (astonzhang) [inline chapter_reinforcement-learning/value-iter.md:98]:
```suggestion
```

File: chapter_reinforcement-learning/index.md [hunk 8]
```diff
@@ -1,8 +1,15 @@
 # Reinforcement Learning
 :label:`chap_reinforcement_learning`
 
+Reinforcement learning (RL) refers to techniques by which an agent learns a policy that optimizes a given performance metric from a sequence of interactions with an environment.  Each of the agent's actions effects the state of the environment which the agent observes. After executing an action, the agent will also receive some scalar value which is called reward. Using this feedback, the agent has to determine which action to take in order to maximize the rewards obtained through the interaction with the environment. By maximizing rewards, the agent can determine the optimal policy. It is worth noting that in contrast to other classes of machine learning methods, like Supervised learning in which a training set has been provided, an RL agent should act in the environment in order to explore its actions' effects and maximize the reward. More importantly, in interactive problems, it is often infeasible to collect data that represent all the cases in which the agent has to act.
+
+![Reinforcement Learning Structure](../img/RL_main.png)
+:width:`400px`
+:label:`fig_rl_big-picture`
```
Comment (astonzhang) [inline chapter_reinforcement-learning/index.md:8]:
```suggestion
```

Per offline discussions, I am removing it to get it built. Please add it back with the figure file later. Please refer to this figure as 
```
:numref:`fig_rl_big`
```
in main text

**Comment by astonzhang** [comment]
* Remove parentheses around :cite:
* Add an empty line right after equations, otherwise:
Screen Shot 2022-12-02 at 11 30 46 AM

See https://github.com/rasoolfa/d2l-en/commit/f0a05b915dced156d42404c53454e82edda2b30b

**Comment by astonzhang** [comment]
* Only one space around operators (e.g., =)
* Capitalize the first letters in code comments
* No docstrings

See 
https://github.com/rasoolfa/d2l-en/commit/5da2392f867056bb58069140e0337268cbefbe15  https://github.com/rasoolfa/d2l-en/commit/2f2a40363a5839dabcdbea26605e666eda25ac20
https://github.com/rasoolfa/d2l-en/blob/vol2-rl/chapter_reinforcement-learning/value-iter.md

File: d2l/torch.py [hunk 9]
```diff
@@ -3071,6 +3071,136 @@ def predict_seq2seq(net, src_sentence, src_vocab, tgt_vocab, num_steps,
         output_seq.append(pred)
     return ' '.join(tgt_vocab.to_tokens(output_seq)), attention_weight_seq
 
+def FrozenLake(seed):
```
[the remark this answers is not in the document]
Comment (astonzhang) [reply d2l/torch.py:3074]: +1
Please revert current changes to torch.py since it's not used in the first two notebooks

File: d2l/torch.py [hunk 10]
```diff
@@ -3071,6 +3071,136 @@ def predict_seq2seq(net, src_sentence, src_vocab, tgt_vocab, num_steps,
...
+            You can only move in the following directions:
+                LEFT which corresponds to action index 0.
+                DOWN which corresponds to action index 1.
+                RIGHT which corresponds to action index 2.
+                UP which corresponds to action index 3.
+            Above text/description is copied from https://gym.openai.com/envs/FrozenLake-v0/
+            and how to process env.P.items is adpated from https://sites.google.com/view/deep-rl-bootcamp/labs
+
+    '''
+    import gym
```
[the remark this answers is not in the document]
Comment (astonzhang) [reply d2l/torch.py:3105]: +1

File: d2l/torch.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (astonzhang) [reply d2l/torch.py:3074]: +1

Please add all these methods to value-iter.md and briefly describe them (no docstring, so you may paraphrase docstrings in main text as descriptions) in main text. Add `#@save` to these functions and run `d2lbook build lib` to make sure they automatically show up in d2l/torch.py.

File: d2l/torch.py [hunk 11]
```diff
@@ -3071,6 +3071,136 @@ def predict_seq2seq(net, src_sentence, src_vocab, tgt_vocab, num_steps,
...
+    env.action_space.seed(seed)
+    env_info = {}
+    env_info['desc'] = env.desc  # 2D array specifying what each grid item means
+    env_info['num_states']     = env.nS # number of observations/states or obs/state dim
+    env_info['num_actions']    = env.nA # number of actions or action dim
+    # define indices for (transition probability, nextstate, reward, done) tuple
+    env_info['trans_prob_idx'] = 0 # index of transition probability entry
+    env_info['nextstate_idx']  = 1 # index of next state entry
+    env_info['reward_idx']     = 2 # index of reward entry
+    env_info['done_idx']       = 3 # index of done entry
```
Comment (astonzhang) [inline d2l/torch.py:3118]: Please apply consistent code style here: 2 spaces before # for inline comments; Capitalize the first letters of code comments; one space around operators.

File: d2l/torch.py [hunk 12]
```diff
@@ -3071,6 +3071,136 @@ def predict_seq2seq(net, src_sentence, src_vocab, tgt_vocab, num_steps,
...
+            # e.g. [(0.3, 0, 0, False), (0.3, 0, 0, False), (0.3, 4, 1, False)]
+            env_info['mdp'][(s,a)] = pxrds
+
+    return env_info
+
+def make_env(name ='', seed=0):
+    """Defined in :numref:`sec_code_rl`"""
+    # Input parameters:
+    # name: specifies a gym environment.
+    # For Value iteration, only FrozenLake-v1 is supported.
```
Comment (astonzhang) [inline d2l/torch.py:3136]: Paraphrase these in main text descriptions

File: d2l/torch.py [hunk 13]
```diff
@@ -3071,6 +3071,136 @@ def predict_seq2seq(net, src_sentence, src_vocab, tgt_vocab, num_steps,
...
+
+    else:
+        raise ValueError("%s env is not supported in this Notebook")
+
+def show_value_function_progress(env_desc, V, pi):
+    """Defined in :numref:`sec_code_rl`"""
+    # This function visualizes how value and policy changes over time.
+    # V:  [num_iters, num_states]
+    # pi: [num_iters, num_states]
+    # How to visualize value function is adapted (but changed) from: https://sites.google.com/view/deep-rl-bootcamp/labs
```
Comment (astonzhang) [inline d2l/torch.py:3148]: Paraphrase these in main text descriptions

File: d2l/torch.py [hunk 14]
```diff
@@ -3071,6 +3071,136 @@ def predict_seq2seq(net, src_sentence, src_vocab, tgt_vocab, num_steps,
...
+
+        # LEFT action: 0.
+        # DOWN action: 1.
+        # RIGHT action: 2.
+        # UP action: 3.
+        action2dxdy = {0: (-.25, 0),
+                       1: (0, .25),
+                       2: (0.25,0),
+                       3: (-.25, 0)
+                      }
```
Comment (astonzhang) [inline d2l/torch.py:3172]: Make these lines more compact. Otherwise we end up with very big code blocks, which are not printing-friendly for books

File: d2l/torch.py [hunk 15]
```diff
@@ -3071,6 +3071,136 @@ def predict_seq2seq(net, src_sentence, src_vocab, tgt_vocab, num_steps,
...
+                    ax.text(x, y, str(env_desc[y,x].decode()),
+                       ha="center", va="center", color="w",
+                         size=20, fontweight='bold')
+
+                else:
+                    ax.text(x, y, str(env_desc[y,x].decode()),
+                       ha="center", va="center", color="g",
+                         size=15, fontweight='bold')
+
+                # no arrow for cells with G and H labels
```
Comment (astonzhang) [inline d2l/torch.py:3194]: Capitalize first letter

**Comment by astonzhang** [comment]
Main issues to address:
* Add source code/imports to md files and run `d2lbook build lib`
* Add back the img file to index.md and refer to it in main text
* Code style fix

File: chapter_reinforcement-learning/value-iter.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (rasoolfa) [reply chapter_reinforcement-learning/value-iter.md]: great.

File: chapter_reinforcement-learning/value-iter.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (rasoolfa) [reply chapter_reinforcement-learning/value-iter.md]: tnx

File: d2l/torch.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (rasoolfa) [reply d2l/torch.py:3074]: yes, those functions are not worth describing in the main text, so I've added them to utils.md

File: d2l/torch.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (rasoolfa) [reply d2l/torch.py:3105]: I've added them to utils.md per @astonzhang's comment.

File: d2l/torch.py (hunk 11 above)
Comment (rasoolfa) [reply d2l/torch.py:3118]: Done.

File: d2l/torch.py (hunk 14 above)
Comment (rasoolfa) [reply d2l/torch.py:3172]: updated. let me know whether or not it looks better!

**Comment by rasoolfa** [comment]
> Last, please create discussion threads under category "English Version - pytorch" (you'll need to sign up an account) at https://discuss.d2l.ai/c/english-version/5
> 
> Example:
> 
> At the end of https://raw.githubusercontent.com/d2l-ai/d2l-en/master/chapter_linear-classification/softmax-regression-scratch.md
> 
> we add (plz remove the topic name in the URL: discuss.d2l.ai/t/[topic name]/[topic id] -> discuss.d2l.ai/t/[topic id])
> 
> ```
> :begin_tab:`pytorch`
> [Discussions](https://discuss.d2l.ai/t/51)
> :end_tab:
> ```
> 
> where https://discuss.d2l.ai/t/softmax-regression-implementation-from-scratch/51 was created

I've added discussion to the value-iter, since my account is on-hold now, I used xx for topic id.

**Comment by AnirudhDagar** [comment]
> I've added discussion to the value-iter, since my account is on-hold now, I used xx for topic id.

Approved on discuss.d2l.ai, should be good now.

**Comment by astonzhang** [comment]
Use transparent background for image next time :) https://github.com/d2l-ai/d2l-en/pull/2361/commits/91d4017cf1dd3b62112c47057f5a7bb6064696f7

**Comment by astonzhang** [comment]
Thanks for contributing the RL chapter! Thanks @AnirudhDagar for the review!

I've made the following touch-ups (you may wish to apply them in your future PRs):

* Use transparent background https://github.com/d2l-ai/d2l-en/pull/2361/commits/91d4017cf1dd3b62112c47057f5a7bb6064696f7
* Capitalize subsection title initials https://github.com/d2l-ai/d2l-en/pull/2361/commits/42bbd6f1c1db2d0b76dc9ef571e18ac30ef2132e
* Add gym import to preface.md to get this import populate to the d2l lib: https://github.com/d2l-ai/d2l-en/pull/2361/commits/48394aa6b6479c9ffe925af45e765f21d1543480
* Run `d2lbook build lib` to populate d2l/*.py with those #@save functions/classes/imports

TODOs for future PRs:
* Add a discussion thread for mdp as well (like what you did for value-iter)
* Add more text/context to the index.md
* (optional) Add more citations/references
